File: library/src/main/java/org/meganekkovr/audio_engine/SoundImpl.java
package org.meganekkovr.audio_engine;

import android.animation.ValueAnimator;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;

import com.google.vr.sdk.audio.GvrAudioEngine;

import java.util.SortedMap;
import java.util.TreeMap;

class SoundImpl implements SoundObject, SoundField, StereoSound {
    private final GvrAudioEngine audioEngine;
    private final int id;
    private final SortedMap<Float, Runnable> timedEvents = new TreeMap<>();
    private float volume = 1;
    private float currentTime;
    private boolean willBeInPlaying;
    private Runnable onEndCallback;

    SoundImpl(GvrAudioEngine audioEngine, int id) {
        this.audioEngine = audioEngine;
        this.id = id;
    }

    @Override
    public void play(boolean loopingEnabled) {
        this.willBeInPlaying = true;
        audioEngine.playSound(id, loopingEnabled);
    }

    @Override
    public void pause() {
        this.willBeInPlaying = false;
        audioEngine.pauseSound(id);
    }

    @Override
    public void resume() {
        this.willBeInPlaying = true;
        audioEngine.resumeSound(id);
    }

    @Override
    public void stop() {
        // I don't set willBeInPlaying to false here for onEndCallback.
        audioEngine.stopSound(id);
    }

    @Override
    public boolean isPlaying() {
        return audioEngine.isSoundPlaying(id);
    }

    @Override
    public void setPosition(float x, float y, float z) {
        audioEngine.setSoundObjectPosition(id, x, y, z);
    }

    @Override
    public void setDistanceRolloffModel(int rolloffModel, float minDistance, float maxDistance) {
        audioEngine.setSoundObjectDistanceRolloffModel(id, rolloffModel, minDistance, maxDistance);
    }

    @Override
    public void setVolume(float volume) {
        this.volume = volume;
        audioEngine.setSoundVolume(id, volume);
    }

    @Override
    public void fadeVolume(float volume, float time) {
        final ValueAnimator animator = ValueAnimator.ofFloat(this.volume, volume)
                .setDuration((long) (time * 1000));

        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                float value = (float) animation.getAnimatedValue();
                SoundImpl.this.setVolume(value);
            }
        });

        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                animator.start();
            }
        });
    }

    @Override
    public void setRotation(float x, float y, float z, float w) {
        audioEngine.setSoundfieldRotation(id, x, y, z, w);
    }

    @Override
    public void onEnd(Runnable callback) {
        this.onEndCallback = callback;
    }

    @Override
    public boolean isEnded() {
        return willBeInPlaying && !isPlaying();
    }

    void notifyOnEnd() {
        if (onEndCallback != null) {
            onEndCallback.run();
            onEndCallback = null;
        }
    }

    @Override
    public void addTimedEvent(float time, @NonNull Runnable event) {
        timedEvents.put(time, event);
    }

    void deltaSeconds(float deltaSeconds) {

        currentTime += deltaSeconds;

        // Emit timed events
        if (!timedEvents.isEmpty()) {
            Float firstKey = timedEvents.firstKey();
            if (firstKey < currentTime) {
                Runnable runnable = timedEvents.remove(firstKey);
                runnable.run();
            }
        }
    }
}

File: library/src/main/java/org/meganekkovr/audio_engine/SoundImpl.kt
package org.meganekkovr.audio_engine

import android.animation.ValueAnimator
import android.os.Handler
import android.os.Looper
import com.google.vr.sdk.audio.GvrAudioEngine
import java.util.*

internal class SoundImpl(private val audioEngine: GvrAudioEngine, private val id: Int) : SoundObject, SoundField, StereoSound {
    private val timedEvents = TreeMap<Float, Runnable>()
    private var volume = 1f
    private var currentTime: Float = 0.toFloat()
    private var willBeInPlaying: Boolean = false
    private var onEndCallback: Runnable? = null

    override val isPlaying: Boolean
        get() = audioEngine.isSoundPlaying(id)

    override val isEnded: Boolean
        get() = willBeInPlaying && !isPlaying

    override fun play(loopingEnabled: Boolean) {
        this.willBeInPlaying = true
        audioEngine.playSound(id, loopingEnabled)
    }

    override fun pause() {
        this.willBeInPlaying = false
        audioEngine.pauseSound(id)
    }

    override fun resume() {
        this.willBeInPlaying = true
        audioEngine.resumeSound(id)
    }

    override fun stop() {
        // I don't set willBeInPlaying to false here for onEndCallback.
        audioEngine.stopSound(id)
    }

    override fun setPosition(x: Float, y: Float, z: Float) {
        audioEngine.setSoundObjectPosition(id, x, y, z)
    }

    override fun setDistanceRolloffModel(rolloffModel: Int, minDistance: Float, maxDistance: Float) {
        audioEngine.setSoundObjectDistanceRolloffModel(id, rolloffModel, minDistance, maxDistance)
    }

    override fun setVolume(volume: Float) {
        this.volume = volume
        audioEngine.setSoundVolume(id, volume)
    }

    override fun fadeVolume(volume: Float, time: Float) {
        val animator = ValueAnimator.ofFloat(this.volume, volume)
                .setDuration((time * 1000).toLong())

        animator.addUpdateListener { animation ->
            val value = animation.animatedValue as Float
            this@SoundImpl.setVolume(value)
        }

        Handler(Looper.getMainLooper()).post { animator.start() }
    }

    override fun setRotation(x: Float, y: Float, z: Float, w: Float) {
        audioEngine.setSoundfieldRotation(id, x, y, z, w)
    }

    override fun onEnd(callback: Runnable) {
        this.onEndCallback = callback
    }

    fun notifyOnEnd() {
        onEndCallback?.run()
        onEndCallback = null
    }

    override fun addTimedEvent(time: Float, event: Runnable) {
        timedEvents[time] = event
    }

    fun deltaSeconds(deltaSeconds: Float) {

        currentTime += deltaSeconds

        // Emit timed events
        if (!timedEvents.isEmpty()) {
            val firstKey = timedEvents.firstKey()
            if (firstKey < currentTime) {
                timedEvents.remove(firstKey)?.run()
            }
        }
    }
}
